### PR TITLE
feat(incidents): implement models and client for /incidents/secrets/

### PR DESCRIPTION
**Pull request by fnareoh**
Implement this endpoint to later display more information in ggshield.

After discussion with Aurélien, we chose to use Litteral when string values are known to be part of limited number of options. We chose not to use Enum in order not to be too restrictive and risk breaking change when a new return value is added in the API.

### CODE REVIEW

File: tests/test_models.py [hunk 1]
```diff
@@ -186,7 +381,10 @@ def test_schema_loads(self, schema_klass, expected_klass, instance_data):
         THEN the extra field should be excluded
         AND the result should be an instance of the expected class
         """
-        schema = schema_klass()
+        if callable(schema_klass):
+            schema = schema_klass()
+        else:
+            schema = schema_klass
```
Comment (agateau-gg) [inline tests/test_models.py:387]: There is some kind of undocumented standard to declare our models and their schema. It looks like this:

```python
@dataclass
class Foo(Base, FromDictMixin):
    # fields


FooSchema = cast(
    Type[BaseSchema], marshmallow_dataclass.class_schema(Foo, BaseSchema)
)

Foo.SCHEMA = FooSchema()
```

If you do this for the schemas you are testing here, you won't need these changes and the code will be a little more consistent.

File: changelog.d/20240823_170420_fnareoh_scrt_4759_optimize_files_to_scan_on_merge_commits.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+### Added
+
+- Added the models and client for to retrieve a secret incident https://api.gitguardian.com/docs#tag/Secret-Incidents/operation/retrieve-incidents
```
Comment (agateau-gg) [inline changelog.d/20240823_170420_fnareoh_scrt_4759_optimize_files_to_scan_on_merge_commits.md:3]:
```suggestion
- Added `GGClient.retrieve_secret_incident()` to retrieve the incident associated with a secret (see https://api.gitguardian.com/docs#tag/Secret-Incidents/operation/retrieve-incidents)
```

File: pygitguardian/client.py [hunk 3]
```diff
@@ -454,6 +455,30 @@ def multi_content_scan(
...
         return obj
 
+    def retrieve_secret_incident(
+        self, incident_id: int, with_occurrences: int = 0
+    ) -> Union[Detail, SecretIncident]:
+        """
+        retrieve_secret_incident handles the /incidents/secret/{incident_id} endpoint of the API
+
+        :param incident_id: incident id
+        :param with_occurrences: number of occurrences of the incident to retrieve (default 0)
```
Comment (agateau-gg) [inline pygitguardian/client.py:465]: Can you clarify whether `0` means no occurrences are retrieved or if it means the  server uses its own default value?

**Review by agateau-gg — CHANGES_REQUESTED** [review]
Minor remarks, but thanks for this not-so-fun work.

File: pygitguardian/models.py [hunk 4]
```diff
@@ -794,3 +794,175 @@ class JWTService(Enum):
...
+    family: str
+    detector_group_name: str
+    detector_group_display_name: str
+
+
+Severity = Literal["low", "medium", "high", "critical", "unknown"]
+ValidityStatus = Literal["valid", "invalid", "failed_to_check", "no_checker", "unknown"]
+IncidentStatus = Literal["IGNORED", "TRIGGERED", "RESOLVED", "ASSIGNED"]
+Tag = Literal[
+    "DEFAULT_BRANCH",
```
Comment (pierrelalanne) [inline pygitguardian/models.py:813]: I don't see the "FALSE_POSITIVE" tag that we recently introduced.  
This is probably missing in some openAPI spec file.

File: tests/test_models.py (hunk 1 above)
Comment (fnareoh) [reply tests/test_models.py]: Hum I was not entirely sure I understood your suggestion but I think successfully applied it ? :sweat_smile:

File: pygitguardian/client.py (hunk 3 above)
Comment (fnareoh) [reply pygitguardian/client.py]: I had put 0 when I didn't think we wanted occurrences but I meant to change this default to 20 (which is the default of the API), done now.

File: pygitguardian/models.py (hunk 4 above)
Comment (fnareoh) [reply pygitguardian/models.py:813]: Yep unfortunately I followed the API Documentation as a start and as Aurélien warned me, it is a fair bit outdated...

**Review by agateau-gg — APPROVED** [review]

**Pull request merged**